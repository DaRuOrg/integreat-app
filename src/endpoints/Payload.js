class Payload {
  constructor (isFetching, data = null, error = null, requestUrl = null) {
    this._isFetching = isFetching
    this._data = data
    this._error = error
    this._requestUrl = requestUrl
  }

  get isFetching () {
    return this._isFetching
  }

  get data () {
    return this._data
  }

  ready () {
    return !!this._data
  }

  get error () {
    return this._error
  }

  get requestUrl () {
    return this._requestUrl
  }
}

export default Payload
