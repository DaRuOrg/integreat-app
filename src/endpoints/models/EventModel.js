export default class EventModel {
  constructor ({ id, title = '', content = '', thumbnail = '', address = '', town = '', date = {} }) {
    this._id = id
    this._title = title
    this._content = content
    this._thumbnail = thumbnail
    this._address = address
    this._town = town
    this._date = date
  }

  get id () {
    return this._id
  }

  get title () {
    return this._title
  }

  get thumbnail () {
    return this._thumbnail
  }

  get content () {
    return this._content
  }

  get address () {
    return this._address + ', ' + this._town
  }

  getDate (locale) {
    return this._date.toLocaleString(locale)
  }
}
