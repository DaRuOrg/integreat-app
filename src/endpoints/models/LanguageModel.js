class LanguageModel {
  constructor (code, name) {
    this._name = code
    this._code = name
  }

  get code () {
    return this._name
  }

  get name () {
    return this._code
  }
}

export default LanguageModel
