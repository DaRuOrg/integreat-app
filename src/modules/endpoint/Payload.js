/**
 * The payload gets stored in the redux store and holds the information about a etch
 */
class Payload {
  constructor (isFetching, data = null, error = null, requestUrl = null, fetchDate = null) {
    this._isFetching = isFetching
    this._fetchDate = fetchDate
    this._data = data
    this._error = error
    this._requestUrl = requestUrl
  }

  /**
   * @return {number} The date the fetch was initiated as serializable number
   */
  get fetchDate () {
    return this._fetchDate
  }

  /**
   * @return {boolean} If a fetch is going on
   */
  get isFetching () {
    return this._isFetching
  }

  /**
   * @return {*} The data which has been fetched or null
   */
  get data () {
    return this._data
  }

  /**
   * @return {boolean} If the {@link data} is ready to be used
   */
  ready () {
    return !!this._data
  }

  /**
   * @return {string} The error message if the fetch failed or null
   */
  get error () {
    return this._error
  }

  /**
   * @return {string} The url which was used to initiate the fetch
   */
  get requestUrl () {
    return this._requestUrl
  }
}

export default Payload
