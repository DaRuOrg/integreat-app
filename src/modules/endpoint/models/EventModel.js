// @flow

import moment from 'moment'

type eventModel = {
  id: boolean,
  title: string,
  content: string,
  thumbnail: string,
  availableLanguages: any,
  content: string,
  town: string,
  address: string,
  startDate: moment,
  endDate: moment,
  allDay: boolean,
  excerpt: string
}

class EventModel {
  _id: boolean
  _title: string
  _content: string
  _thumbnail: string
  _availableLanguages: any
  _content: string
  _town: string
  _address: string
  _startDate: moment
  _endDate: moment
  _allDay: boolean
  _excerpt: string

  constructor (eventModel: eventModel) {
    this._id = eventModel.id
    this._title = eventModel.title
    this._content = eventModel.content
    this._thumbnail = eventModel.thumbnail
    this._address = eventModel.address
    this._town = eventModel.town
    this._startDate = eventModel.startDate
    this._endDate = eventModel.endDate
    this._allDay = eventModel.allDay
    this._excerpt = eventModel.excerpt
    this._availableLanguages = eventModel.availableLanguages
  }

  get id (): boolean {
    return this._id
  }

  get title (): string {
    return this._title
  }

  get thumbnail (): string {
    return this._thumbnail
  }

  get content (): string {
    return this._content
  }

  get address (): ?string {
    if (this._address && this._town) {
      return `${this._address}, ${this._town}`
    } else if (this._address) {
      return this._address
    } else if (this._town) {
      return this._town
    }
  }

  get startDate (): moment {
    return this._startDate
  }

  get endDate (): moment {
    return this._endDate
  }

  get allDay (): boolean {
    return this._allDay
  }

  get excerpt (): string {
    return this._excerpt
  }

  get availableLanguages (): any {
    return this._availableLanguages
  }
}

export default EventModel
