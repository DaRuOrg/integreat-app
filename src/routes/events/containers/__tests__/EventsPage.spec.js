import React from 'react'
import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import ConnectedEventsPage, { EventsPage } from '../EventsPage'
import EventModel from 'modules/endpoint/models/EventModel'
import LanguageModel from 'modules/endpoint/models/LanguageModel'
import Payload from '../../../../modules/endpoint/Payload'

describe('EventsPage', () => {
  const events = [
    new EventModel({
      id: 1234,
      title: 'first Event',
      availableLanguages: {de: '1235', ar: '1236'}
    }),
    new EventModel({
      id: 1235,
      title: 'erstes Event',
      availableLanguages: {en: '1234', ar: '1236'}
    }),
    new EventModel({
      id: 2,
      title: 'second Event'
    })
  ]

  const location = 'augsburg'
  const languages = [
    new LanguageModel('en', 'English'),
    new LanguageModel('de', 'Deutsch'),
    new LanguageModel('ar', 'Arabic')
  ]
  const language = 'en'
  const id = '1235'

  test('should render EventList', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const wrapper = shallow(
      <EventsPage events={events}
                  location={location}
                  languages={languages}
                  language={language}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('should render EventDetail', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const wrapper = shallow(
      <EventsPage events={events}
                  location={location}
                  languages={languages}
                  language={language}
                  id={id}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('should render Spinner', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const wrapper = shallow(
      <EventsPage events={[]}
                  location={location}
                  languages={languages}
                  language={language}
                  id={id}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('should dispatch once on mount with availableLanguages', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const eventsPage = shallow(
      <EventsPage events={events}
                  location={location}
                  languages={languages}
                  language={language}
                  id={id}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    ).instance()

    expect(mockDispatchLanguageChangeUrls.mock.calls).toHaveLength(1)
    expect(mockDispatchLanguageChangeUrls).toBeCalledWith(
      eventsPage.mapLanguageToUrl, languages, events[1].availableLanguages
    )
  })

  test('should dispatch once on mount without availableLanguages', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const eventsPage = shallow(
      <EventsPage events={events}
                  location={location}
                  languages={languages}
                  language={language}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    ).instance()

    expect(mockDispatchLanguageChangeUrls.mock.calls).toHaveLength(1)
    expect(mockDispatchLanguageChangeUrls).toBeCalledWith(eventsPage.mapLanguageToUrl, languages, {})
  })

  test('should dispatch on prop update with availableLanguages', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const wrapper = shallow(
      <EventsPage events={[]}
                  location={location}
                  languages={languages}
                  language={language}
                  id={id}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    )

    expect(mockDispatchLanguageChangeUrls.mock.calls).toHaveLength(1)

    wrapper.setProps({events: events, ...wrapper.props})

    expect(mockDispatchLanguageChangeUrls.mock.calls).toHaveLength(2)
    expect(mockDispatchLanguageChangeUrls).toBeCalledWith(
      wrapper.instance().mapLanguageToUrl, languages, events[1].availableLanguages
    )
  })

  test('should not dispatch on prop update', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const wrapper = shallow(
      <EventsPage events={events}
                  location={location}
                  languages={languages}
                  language={language}
                  id={id}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    )

    let mockCalls = mockDispatchLanguageChangeUrls.mock.calls

    wrapper.setProps({events: events, ...wrapper.props})

    expect(mockDispatchLanguageChangeUrls.mock.calls).toHaveLength(mockCalls.length)

    wrapper.setProps({...wrapper.props})

    expect(mockDispatchLanguageChangeUrls.mock.calls).toHaveLength(mockCalls.length)
  })

  test('mapLanguageToUrl', () => {
    const mockDispatchLanguageChangeUrls = jest.fn()

    const mapLanguageToUrl = shallow(
      <EventsPage events={events}
                  location={location}
                  languages={languages}
                  language={language}
                  id={id}
                  dispatchLanguageChangeUrls={mockDispatchLanguageChangeUrls} />
    ).instance().mapLanguageToUrl

    expect(mapLanguageToUrl('en')).toBe('/augsburg/en/events')
    expect(mapLanguageToUrl('en', 1234)).toBe('/augsburg/en/events/1234')
  })

  const mockStore = configureMockStore([thunk])

  describe('connect', () => {
    test('should map state to props', () => {
      const store = mockStore({
        events: new Payload(false),
        languages: new Payload(false),
        router: {params: {location: 'augsburg', language: 'en', id: '1234'}}
      })

      const tree = mount(
        <Provider store={store}>
          <ConnectedEventsPage />
        </Provider>
      )

      const eventsPageProps = tree.find(ConnectedEventsPage).childAt(0).props()

      expect(eventsPageProps).toEqual({
        location: 'augsburg',
        language: 'en',
        id: '1234',
        events: events,
        languages: languages,
        dispatchLanguageChangeUrls: expect.any(Function)
      })
    })

    test('should map dispatch to props', () => {
      const store = mockStore({
        events: new Payload(false),
        languages: new Payload(false),
        router: {params: {location: location, language: language}}
      })

      const mapLanguageToUrl = (language, id) => 'test' + language + id

      const testUrls = {
        en: 'testenundefined',
        de: 'testde1235',
        ar: 'testar1236'
      }

      const availableLanguages = {
        de: '1235',
        ar: '1236'
      }

      const tree = mount(
        <Provider store={store}>
          <ConnectedEventsPage />
        </Provider>
      )

      // todo expect setLanguageChangeUrls action to be in store, but as we don't get events and languages from our
      // mocked endpoint no action is dispatched

      const eventsPageProps = tree.find(ConnectedEventsPage).childAt(0).props()

      let countActions = store.getActions().length

      eventsPageProps.dispatchLanguageChangeUrls(mapLanguageToUrl, languages, availableLanguages)
      expect(store.getActions()).toHaveLength(countActions + 1)

      expect(store.getActions()).toContainEqual({
        payload: testUrls,
        type: 'SET_LANGUAGE_CHANGE_URLS'
      })
    })
  })
})
