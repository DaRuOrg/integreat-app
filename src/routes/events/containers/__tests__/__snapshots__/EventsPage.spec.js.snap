// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsPage should render EventDetail 1`] = `
<Component
  event={
    EventModel {
      "_address": "",
      "_availableLanguages": Object {
        "ar": "1236",
        "en": "1234",
      },
      "_content": "",
      "_dateModel": DateModel {
        "_allDay": undefined,
        "_endDate": undefined,
        "_startDate": undefined,
      },
      "_excerpt": "",
      "_id": 1235,
      "_thumbnail": "",
      "_title": "erstes Event",
      "_town": "",
    }
  }
  language="en"
  location="augsburg"
/>
`;

exports[`EventsPage should render EventList 1`] = `
<Component
  events={
    Array [
      EventModel {
        "_address": "",
        "_availableLanguages": Object {
          "ar": "1236",
          "de": "1235",
        },
        "_content": "",
        "_dateModel": DateModel {
          "_allDay": undefined,
          "_endDate": undefined,
          "_startDate": undefined,
        },
        "_excerpt": "",
        "_id": 1234,
        "_thumbnail": "",
        "_title": "first Event",
        "_town": "",
      },
      EventModel {
        "_address": "",
        "_availableLanguages": Object {
          "ar": "1236",
          "en": "1234",
        },
        "_content": "",
        "_dateModel": DateModel {
          "_allDay": undefined,
          "_endDate": undefined,
          "_startDate": undefined,
        },
        "_excerpt": "",
        "_id": 1235,
        "_thumbnail": "",
        "_title": "erstes Event",
        "_town": "",
      },
      EventModel {
        "_address": "",
        "_availableLanguages": Array [],
        "_content": "",
        "_dateModel": DateModel {
          "_allDay": undefined,
          "_endDate": undefined,
          "_startDate": undefined,
        },
        "_excerpt": "",
        "_id": 2,
        "_thumbnail": "",
        "_title": "second Event",
        "_town": "",
      },
    ]
  }
  language="en"
  url="/augsburg/en/events"
/>
`;

exports[`EventsPage should render Spinner 1`] = `
<Spinner
  fadeIn="full"
  name="line-scale-party"
  noFadeIn={false}
  overrideSpinnerClassName=""
/>
`;
